### DISCUSSION TITLE
Tackling memory leaks

### DISCUSSION THREAD

**Discussion by kamakazikamikaze** [easysnmp/easysnmp#139] [General]
I've created branch [memork-leaks](/kamakazikamikaze/easysnmp/tree/memory-leaks) to address outstanding issues regarding memory leaks. The first commit appears to reduce leaks from `13,684 bytes in 261 blocks` to `10,044 bytes in 248 blocks`. It's not much but it's a start. The next chunk of leaks I need to tackle seem to originate from v3 sessions:

```sh
62 bytes in 62 blocks are definitely lost in loss record 600 of 786
   at 0x483DD99: calloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x8774B25: hex_to_binary2 (tools.c:470)
   by 0x86E2ED2: netsnmp_create_session_v3 (interface.c:1693)
   by 0x3E2546: cfunction_call (methodobject.c:552)
   by 0x179297: _PyObject_MakeTpCall (call.c:215)
   by 0x26935E: _PyObject_VectorcallTstate (abstract.h:112)
   by 0x2693D9: PyObject_Vectorcall (abstract.h:123)
   by 0x2824B5: trace_call_function (ceval.c:5822)
   by 0x28280A: call_function (ceval.c:5864)
   by 0x27C31A: _PyEval_EvalFrameDefault (ceval.c:4181)
   by 0x269679: _PyEval_EvalFrame (pycore_ceval.h:46)
   by 0x280714: _PyEval_Vector (ceval.c:5065)

62 bytes in 62 blocks are definitely lost in loss record 601 of 786
   at 0x483DD99: calloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x8774B25: hex_to_binary2 (tools.c:470)
   by 0x86E2F20: netsnmp_create_session_v3 (interface.c:1696)
   by 0x3E2546: cfunction_call (methodobject.c:552)
   by 0x179297: _PyObject_MakeTpCall (call.c:215)
   by 0x26935E: _PyObject_VectorcallTstate (abstract.h:112)
   by 0x2693D9: PyObject_Vectorcall (abstract.h:123)
   by 0x2824B5: trace_call_function (ceval.c:5822)
   by 0x28280A: call_function (ceval.c:5864)
   by 0x27C31A: _PyEval_EvalFrameDefault (ceval.c:4181)
   by 0x269679: _PyEval_EvalFrame (pycore_ceval.h:46)
   by 0x280714: _PyEval_Vector (ceval.c:5065)

4,960 bytes in 62 blocks are definitely lost in loss record 744 of 786
   at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x8759934: snmp_duplicate_objid (snmp_api.c:7909)
   by 0x86E301A: netsnmp_create_session_v3 (interface.c:1722)
   by 0x3E2546: cfunction_call (methodobject.c:552)
   by 0x179297: _PyObject_MakeTpCall (call.c:215)
   by 0x26935E: _PyObject_VectorcallTstate (abstract.h:112)
   by 0x2693D9: PyObject_Vectorcall (abstract.h:123)
   by 0x2824B5: trace_call_function (ceval.c:5822)
   by 0x28280A: call_function (ceval.c:5864)
   by 0x27C31A: _PyEval_EvalFrameDefault (ceval.c:4181)
   by 0x269679: _PyEval_EvalFrame (pycore_ceval.h:46)
   by 0x280714: _PyEval_Vector (ceval.c:5065)

4,960 bytes in 62 blocks are definitely lost in loss record 745 of 786
   at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x8759934: snmp_duplicate_objid (snmp_api.c:7909)
   by 0x86E31ED: netsnmp_create_session_v3 (interface.c:1769)
   by 0x3E2546: cfunction_call (methodobject.c:552)
   by 0x179297: _PyObject_MakeTpCall (call.c:215)
   by 0x26935E: _PyObject_VectorcallTstate (abstract.h:112)
   by 0x2693D9: PyObject_Vectorcall (abstract.h:123)
   by 0x2824B5: trace_call_function (ceval.c:5822)
   by 0x28280A: call_function (ceval.c:5864)
   by 0x27C31A: _PyEval_EvalFrameDefault (ceval.c:4181)
   by 0x269679: _PyEval_EvalFrame (pycore_ceval.h:46)
   by 0x280714: _PyEval_Vector (ceval.c:5065)
```

There are a *lot* of "possibly lost" blocks that do not seem to tie directly back to any invoking code from interface.c, so for now I'm going to assume that these are caused from strings created in Python that get passed to our v3 sessions and are lost because they are not able to be safely deallocated. So hopefully tackling the "definitely lost" results in clearing the "possibly lost".

**Comment by carlkidcrypto** [comment]
Glad to hear work is being done to address the memory leaks. For reference to myself and others. Here is the memory leak I see on 0.2.5 on Ubuntu 20. I have a python script that runs snmp queries repeatedly once every second on 16 snmp enabled devices. I had to write a systemd service file that would restart my logger once every 4 hours. Otherwise the memory profile grows significantly. I'll swap over to the memory leak branch and see what the difference is.
![image](https://user-images.githubusercontent.com/27721404/158414021-7973ace3-6ac4-4324-b61e-9eecb04b3e7b.png)

Update: Switched to the memory leak branch. Now we wait and see 👍🏽

**Reply by kamakazikamikaze** [reply]
I don't expect there to be much saved at the moment but I am curious to see what kind of a difference the one change has made. I've got a hunch that I'll need to change the session initiation to instead call `snmp_open` as the Net-SNMP library uses this for the client applications.

**Reply by kamakazikamikaze** [reply]
@carlkid1499 Could you share what OIDs you're polling and what version (v1, v2, v3) you use for polling?

**Reply by carlkidcrypto** [reply]
As expected, there wasn't much change in the memory profile.
![image](https://user-images.githubusercontent.com/27721404/158630185-99e12a64-bc48-4880-ab59-f4e169dfdbeb.png)

The devices I poll with my snmp logger use 'v2c'. They use a community password since the devices are read only.
This snippet is how the sessions are made:
```python
       # Create a fresh SNMP session
        session = easysnmp.Session(hostname=device_ip,
                                   community="PASSWORD_HERE", version=2)
```

This snippet is how OID's are fetched from the devices:
```python
        # Use the session to get what we want
        for key, val in self.__OID_dict.items():
            # We add a try block here since it can fail if a device isn't available
            try:
                results = session.get(val)
                if self.__debug:
                    print("Getting:", key, results)
                    print(results.value)

            except SystemError:
                # If we get here than the device doesn't have snmp enabled
                # Error looks like this: <built-in function get> returned NULL without setting an error
                if self.__debug:
                    print("Device doesn't have snmp, continue anyway!")

                # break loop and move onto next device. No use in trying next item
                break

        # Close our SNMP session
        del session
```

These are the OIDs:
```python
        # Define the names for the OID's in a dictionary
        self.__OID_dict = {"NAME_FOR_OID_HERE":  "iso.3.6.1.4.1.31823.1.2488.2.7.1.1.0",
                           "NAME_FOR_OID_HERE": "iso.3.6.1.4.1.31823.1.2488.2.7.1.2.0",
                           "NAME_FOR_OID_HERE":  "iso.3.6.1.4.1.31823.1.2488.2.7.1.3.0",
                           "NAME_FOR_OID_HERE": "iso.3.6.1.4.1.31823.1.2488.2.3.1.4.1.3.1",
                           "NAME_FOR_OID_HERE": "iso.3.6.1.4.1.31823.1.2488.2.3.1.4.1.3.2",
                           "NAME_FOR_OID_HERE": "iso.3.6.1.4.1.31823.1.2488.2.3.1.4.1.4.1",
                           "NAME_FOR_OID_HERE": "iso.3.6.1.4.1.31823.1.2488.2.3.1.4.1.4.2",
                           "NAME_FOR_OID_HERE": "iso.3.6.1.4.1.31823.1.2488.2.7.1.4.0",
                           "NAME_FOR_OID_HERE": "iso.3.6.1.4.1.31823.1.2488.2.7.1.5.0",
                           "NAME_FOR_OID_HERE": "iso.3.6.1.4.1.31823.1.2488.2.7.1.6.0",
                           "NAME_FOR_OID_HERE": "iso.3.6.1.4.1.31823.1.2488.2.7.1.7.0"}
```

**Reply by kamakazikamikaze** [reply]
Looks like your call to query SNMP is within a function scope and exits from that when done polling. I'm assuming you've got something within Python that controls timing to poll at a specified interval or time. If that's the case, and assuming you're using Python 3, the garbage collector should be releasing strings from results once the function/for-loop exit. If it isn't, then our issue may also be linked to response values.

**Reply by carlkidcrypto** [reply]
Yup, the creation, usage, and deletion for the SNMP sessions are all within a functions scope. Correct, I have something that controls the poll time. To be precise I am using Python3 (version 3.8.10). The control loop goes like this:
1. run the get_snmp_data function (snippets shown above)
2. do stuff with the data i.e insert into database, textfile, etc
3. sleep 1
4. do it all over again.
```python
results = session.get(val)
```
True, the results should be cleared by the garbage collector eventually. If that was the case then we wouldn't see the memory profile grow. Like you said if we assume the garbage collector isn't removing the results, then the issue may be related to: what is/should be happening when results goes out of scope.

Here results is an 'SNMPVariable type':
![image](https://user-images.githubusercontent.com/27721404/158661890-316c1e5d-e02c-4565-ba39-6755b33bccfc.png)

Perhaps we need to define what cleanup happens when we ```__enter__``` scope and ```__exit__``` scope ?
https://docs.python.org/3/library/contextlib.html#cleaning-up-in-an-enter-implementation

**Reply by carlkidcrypto** [reply]
On a separate note, let me try explicitly deleting my 'results' just before scope is exited. So rather than
just this:
```python
# Close our SNMP session
del session
```

I'll do this:
```python
# Close our SNMP session
del session
del results
```

**Reply by kamakazikamikaze** [reply]
> On a separate note, let me try explicitly deleting my 'results' just before scope is exited. So rather than
> 
> just this:
> 
> ```python
> 
> # Close our SNMP session
> 
> del session
> 
> ```
> 
> 
> 
> I'll do this:
> 
> ```python
> 
> # Close our SNMP session
> 
> del session
> 
> del results
> 
> ```
> 
> 

You can also try manually invoking the garbage collector with
```py
import gc

# Line after your function call
gc.collect()
```
to see if that will forcefully release some memory back. But again, Python should already be doing that and the growth in memory consumption shows it's not able to fully release it.

**Reply by kamakazikamikaze** [reply]
> True, the results should be cleared by the garbage collector eventually. If that was the case then we wouldn't see the memory profile grow. Like you said if we assume the garbage collector isn't removing the results, then the issue may be related to: what is/should be happening when results goes out of scope.
> 
> 
> 
> Here results is an 'SNMPVariable type':
> 
> ![image](https://user-images.githubusercontent.com/27721404/158661890-316c1e5d-e02c-4565-ba39-6755b33bccfc.png)
> 
> 
> 
> Perhaps we need to define what cleanup happens when we ```__enter__``` scope and ```__exit__``` scope ?
> 
> https://docs.python.org/3/library/contextlib.html#cleaning-up-in-an-enter-implementation
> 

Yup, I'm going to look at the variable results after I figure out how to handle the v3 session memory leaks. I think the result strings get set on the C side so it's possible that we are mishandling something.

**Reply by carlkidcrypto** [reply]
> > On a separate note, let me try explicitly deleting my 'results' just before scope is exited. So rather than
> > just this:
> > ```python
> > # Close our SNMP session
> > 
> > del session
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'll do this:
> > ```python
> > # Close our SNMP session
> > 
> > del session
> > 
> > del results
> > ```
> 
> You can also try manually invoking the garbage collector with
> 
> ```python
> import gc
> 
> # Line after your function call
> gc.collect()
> ```
> 
> to see if that will forcefully release some memory back. But again, Python should already be doing that and the growth in memory consumption shows it's not able to fully release it.

Added the extra delete first by itself which resulted in no change to the memory profile. Then I only added the manual call to the gc as suggested. still the same memory profile:
![image](https://user-images.githubusercontent.com/27721404/158836737-378722cd-2cca-42e3-95d4-ee750bd4f00a.png)

**Reply by kamakazikamikaze** [reply]
Thanks for the thorough testing. I'll balance out my time for both SNMPVariables and v3 sessions.

**Reply by kamakazikamikaze** [reply]
I'm replicating the test that I asked you to perform and my results are a bit different. Granted, I'm using Python 3.10.2 with debug symbols and running with `PYMALLOC=malloc` for tracing purposes, so there may be some discrepancies due to the use of a different malloc function.

My code is simple:

```py
from easysnmp import Session
from gc import collect
from time import sleep

SESS_ARGS_V3 = {
    'version': 3,
    'hostname': 'localhost',
    'remote_port': 11161,
    'security_level': 'authPriv',
    'security_username': 'initial',
    'privacy_password': 'priv_pass',
    'auth_password': 'auth_pass'
}


if __name__ == '__main__':
    sess = Session(**SESS_V3_ARGS)
    for i in range(10):
        res = sess.walk('.')
        del res
        collect()
        sleep(5)
```

Invoked with `PYMALLOC=malloc python -m palanteer simple_bulkwalk.py`

What I'm seeing is that memory usage returns to a stable level after each run. Monitoring memory in `htop`, I see that each loop returns to roughly 131M VIRT/37660 RES/10800 SHR by the time it sleeps. So the `SNMPVariable` is probably not to blame. I'm going to modify the script to have different testing scenarios, one of which will create and destroy a `Session` instance each loop like what you're doing. Hopefully that will help narrow things down.
<img width="1245" alt="image" src="https://user-images.githubusercontent.com/8862823/159041749-71ebca64-8f07-470c-b856-54f44a2c8e6b.png">

**Reply by carlkidcrypto** [reply]
> I'm replicating the test that I asked you to perform and my results are a bit different. Granted, I'm using Python 3.10.2 with debug symbols and running with `PYMALLOC=malloc` for tracing purposes, so there may be some discrepancies due to the use of a different malloc function.
> 
> My code is simple:
> 
> ```python
> from easysnmp import Session
> from gc import collect
> from time import sleep
> 
> SESS_ARGS_V3 = {
>     'version': 3,
>     'hostname': 'localhost',
>     'remote_port': 11161,
>     'security_level': 'authPriv',
>     'security_username': 'initial',
>     'privacy_password': 'priv_pass',
>     'auth_password': 'auth_pass'
> }
> 
> 
> if __name__ == '__main__':
>     sess = Session(**SESS_V3_ARGS)
>     for i in range(10):
>         res = sess.walk('.')
>         del res
>         collect()
>         sleep(5)
> ```
> 
> Invoked with `PYMALLOC=malloc python -m palanteer simple_bulkwalk.py`
> 
> What I'm seeing is that memory usage returns to a stable level after each run. Monitoring memory in `htop`, I see that each loop returns to roughly 131M VIRT/37660 RES/10800 SHR by the time it sleeps. So the `SNMPVariable` is probably not to blame. I'm going to modify the script to have different testing scenarios, one of which will create and destroy a `Session` instance each loop like what you're doing. Hopefully that will help narrow things down. <img alt="image" width="1245" src="https://user-images.githubusercontent.com/8862823/159041749-71ebca64-8f07-470c-b856-54f44a2c8e6b.png">

Well that sure is interesting. I also notice you are using SNMP V3. Perhaps try running V3 and V2 side by side ? That's the only difference between mine and your tests besides python versions.

**Reply by kamakazikamikaze** [reply]
I'll give that a go. It could also be that `palanteer` isn't interpreting memory properly, or that I need to run on Python 3.8.10 to get the same results.

For reference, my "finalized" script is as follows:

```py
from argparse import ArgumentParser
from easysnmp import Session
from gc import collect
from time import sleep

SESS_ARGS = {
    '1': {
        'version': 1,
        'hostname': 'localhost',
        'remote_port': 11161,
        'community': 'public'
    },
    '2': {
        'version': 2,
        'hostname': 'localhost',
        'remote_port': 11161,
        'community': 'public'
    },
    '3': {
        'version': 3,
        'hostname': 'localhost',
        'remote_port': 11161,
        'security_level': 'authPriv',
        'security_username': 'initial',
        'privacy_password': 'priv_pass',
        'auth_password': 'auth_pass'
    }
}


def bloat_memory(test, query, method, version, loops, pause, no_collect=False):
    if test == 'persist':
        sess = Session(**SESS_ARGS[version])
        m = getattr(sess, method)
    for i in range(loops):
        if test == 'new':
            sess = Session(**SESS_ARGS[version])
            m = getattr(sess, method)
        res = m(query)
        del res
        if test == 'new':
            del m
            del sess
        if not no_collect:
            collect()
        if pause:
            sleep(pause)


if __name__ == '__main__':
    agp = ArgumentParser()
    agp.add_argument('test', choices=('persist', 'new'))
    agp.add_argument('method', choices=('get', 'walk', 'bulkwalk'))
    agp.add_argument('version', choices=('1', '2', '3'))
    agp.add_argument('-l', '--loops', default=10, type=int)
    agp.add_argument('-p', '--pause', default=5, type=int)
    agp.add_argument('-q', '--query', default='.', type=str)
    agp.add_argument('--no-gc', action='store_true')
    args = agp.parse_args()

    bloat_memory(args.test, args.query, args.method, args.version, args.loops, args.pause, args.no_gc)
```

**Reply by kamakazikamikaze** [reply]
Ok, I've made a few changes to the Valgrind tracing to filter down to more probable leaks and there's a decent list of things for me to look into.

Ran: `PYTHONMALLOC=malloc valgrind --tool=memcheck --trace-children=yes --leak-check=full --show-leak-kinds=all --suppressions=dev/valgrind-python.supp --log-file=dev/valgrind_$(date +%y%m%d_%H%M%S).log python dev/test_different_methods.py new walk 2 -p 1 -l 15`

Snippet:

```
==531809== 540 bytes in 15 blocks are definitely lost in loss record 1,654 of 1,824
==531809==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==531809==    by 0x1C949A: _PyMem_RawMalloc (obmalloc.c:99)
==531809==    by 0x1CA689: PyObject_Malloc (obmalloc.c:685)
==531809==    by 0x16743E: _PyBytes_FromSize (bytesobject.c:126)
==531809==    by 0x16753E: PyBytes_FromStringAndSize (bytesobject.c:159)
==531809==    by 0x22A7CF: _PyUnicode_AsLatin1String (unicodeobject.c:7381)
==531809==    by 0x21964D: PyUnicode_AsEncodedString (unicodeobject.c:3946)
==531809==    by 0x65495AE: py_netsnmp_attr_string (interface.c:1357)
==531809==    by 0x654C3E2: netsnmp_walk (interface.c:2721)
==531809==    by 0x3E2546: cfunction_call (methodobject.c:552)
==531809==    by 0x179297: _PyObject_MakeTpCall (call.c:215)
==531809==    by 0x26935E: _PyObject_VectorcallTstate (abstract.h:112)
...
==531809== LEAK SUMMARY:
==531809==    definitely lost: 540 bytes in 15 blocks
==531809==    indirectly lost: 0 bytes in 0 blocks
==531809==      possibly lost: 521,546 bytes in 5,309 blocks
==531809==    still reachable: 719,249 bytes in 13,129 blocks
==531809==         suppressed: 0 bytes in 0 blocks
```

I'll do some checks on the Net-SNMP library/CLI tools themselves to see what types of memory issues are reported there.

**Reply by kamakazikamikaze** [reply]
`snmpbulkwalk`:
```
==532866== LEAK SUMMARY:
==532866==    definitely lost: 7 bytes in 1 blocks
==532866==    indirectly lost: 0 bytes in 0 blocks
==532866==      possibly lost: 0 bytes in 0 blocks
==532866==    still reachable: 701,486 bytes in 13,067 blocks
==532866==         suppressed: 0 bytes in 0 blocks
```

`snmpwalk`:
```
==532899== LEAK SUMMARY:
==532899==    definitely lost: 7 bytes in 1 blocks
==532899==    indirectly lost: 0 bytes in 0 blocks
==532899==      possibly lost: 0 bytes in 0 blocks
==532899==    still reachable: 701,943 bytes in 13,085 blocks
==532899==         suppressed: 0 bytes in 0 blocks
```

`easysnmp.Session.get`:
```
==533338== LEAK SUMMARY:
==533338==    definitely lost: 299 bytes in 6 blocks
==533338==    indirectly lost: 0 bytes in 0 blocks
==533338==      possibly lost: 517,206 bytes in 5,296 blocks
==533338==    still reachable: 719,249 bytes in 13,129 blocks
==533338==         suppressed: 0 bytes in 0 blocks
```

So we have more leaks than I previously understood. Alright, time to get digging!

**Reply by carlkidcrypto** [reply]
Okay, sweet! Looks like we were able to narrow those memory leaks down! Give me a mention when you have stuff pushed to the memory leak branch. I can help with testing and verification 👍🏽

**Reply by carlkidcrypto** [reply]
Looks like Python just had a maintenance release not too long ago: https://www.python.org/downloads/release/python-3103/
Interesting enough there was a memory leak fix in 3.10.3: 
![image](https://user-images.githubusercontent.com/27721404/159084243-8c843065-691a-47a2-b470-9221435aab1e.png)

I wonder if this might be adding to the EasySNMP memory leak hmmm....

**Reply by kamakazikamikaze** [reply]
Interesting, but following the bug report discussion it seems that it was a 2021 patch that refactored the code and caused the leaks.

https://github.com/python/cpython/commit/0332e569c12d3dc97171546c6dc10e42c27de34b

I don't think it's our culprit.

I did find, however, that `py_netsnmp_attr_string` has a bug. We're converting an attribute value into bytes and pointing to the memory space, but we don't handle decref'ing the object that is returned. Attempting to decref it, however, causes a read-after-free warning in Valgrind. So there's probably going to be a bit of refactoring done to properly maintain and decref the object when we're done with it.

**Reply by kamakazikamikaze** [reply]
Created #141 to track

**Reply by kamakazikamikaze** [reply]
Alright, pushed a0ef2e to `memory-leaks`. Check it out and see if it helps!

**Reply by kamakazikamikaze** [reply]
Also, @carlkid1499 you could probably reduce memory consumption by holding on to the sessions between runs. Each session appears to re-initialize the SNMP library which doesn't necessarily leak memory but may contribute to the residual size. If you'd like assistance in refactoring to make this work I'd be more than happy to help provide some examples.

**Reply by carlkidcrypto** [reply]
> Also, @carlkid1499 you could probably reduce memory consumption by holding on to the sessions between runs. Each session appears to re-initialize the SNMP library which doesn't necessarily leak memory but may contribute to the residual size. If you'd like assistance in refactoring to make this work I'd be more than happy to help provide some examples.

Thanks for doing all that work. Just now pulling the changes down for some tests. Yeah, I could get by with not deleting the session. It's being deleted since 16 threads are launched to collect snmp stats from all 16 devices simultaneously. To be safe and thread safe I make and destroy each session within the threads. So really the logic is like:

1. pass parameters to thread func (not the easysnmp session) i.e IP of device.
2. each thread executes making one session and then fetching several distinct OIDs
3. each thread ends with deleting its session.
4. main thread joins all threads back
5. Sleep and then repeat.

**Reply by kamakazikamikaze** [reply]
> Thanks for doing all that work. Just now pulling the changes down for some tests. Yeah, I could get by with not deleting the session. It's being deleted since 16 threads are launched to collect snmp stats from all 16 devices simultaneously. To be safe and thread safe I make and destroy each session within the threads. So really the logic is like:
> 
> 
> 
> 1. pass parameters to thread func (not the easysnmp session) i.e IP of device.
> 
> 2. each thread executes making one session and then fetching several distinct OIDs
> 
> 3. each thread ends with deleting its session.
> 
> 4. main thread joins all threads back
> 
> 5. Sleep and then repeat. 

Oof, yeah ok I understand now why you're doing that. Way back when I used this library for work, we had to spawn multiple processes and assign one target device per process, then set a limit of max tasks per process so memory could be free'd back to the system. (We were querying 3000+ devices per run every 15 minutes.) I don't think you'd want to move in that direction.

Maybe I can see about making a session manager that initializes the library and tracks sessions, that way you can have a thread-safe way to query multiple devices and not need to destroy sessions repeatedly. No guarantees tho :/

**Reply by carlkidcrypto** [reply]
> > Thanks for doing all that work. Just now pulling the changes down for some tests. Yeah, I could get by with not deleting the session. It's being deleted since 16 threads are launched to collect snmp stats from all 16 devices simultaneously. To be safe and thread safe I make and destroy each session within the threads. So really the logic is like:
> > 
> > 1. pass parameters to thread func (not the easysnmp session) i.e IP of device.
> > 2. each thread executes making one session and then fetching several distinct OIDs
> > 3. each thread ends with deleting its session.
> > 4. main thread joins all threads back
> > 5. Sleep and then repeat.
> 
> Oof, yeah ok I understand now why you're doing that. Way back when I used this library for work, we had to spawn multiple processes and assign one target device per process, then set a limit of max tasks per process so memory could be free'd back to the system. (We were querying 3000+ devices per run every 15 minutes.) I don't think you'd want to move in that direction.
> 
> Maybe I can see about making a session manager that initializes the library and tracks sessions, that way you can have a thread-safe way to query multiple devices and not need to destroy sessions repeatedly. No guarantees tho :/

Wow 3000+ devices that's a huge work load!. For sure, for now it works as is. Making a session manager would be awesome, but IMHO not needed as of right now. Plus debugging session managers...... well I'll leave it at that lol. 

But seriously, thank you for all this work to find the memory leaks!

**Comment by kamakazikamikaze** [comment]
My valgrind output definitely appears to only have issues with v3 sessions. Commenting out SNMPv3 session arguments shows zero `definitely lost` blocks.

It's honestly strange as I comb through the code. Referencing the above blocks:

```c
/* lines 1692 to 1697, for blocks 600 to 601 */
    session.securityEngineIDLen =
        hex_to_binary2((unsigned char *)sec_eng_id, STRLEN(sec_eng_id),
                       (char **) &session.securityEngineID);
    session.contextEngineIDLen =
        hex_to_binary2((unsigned char *)context_eng_id, STRLEN(sec_eng_id),
                       (char **) &session.contextEngineID);

/* lines 1719 to 1723, for block 744 */
    else if (!strcmp(auth_proto, "DEFAULT"))
    {
        const oid* a = get_default_authtype(&session.securityAuthProtoLen);
        session.securityAuthProto
            = snmp_duplicate_objid(a, session.securityAuthProtoLen);
    }

/* lines 1765 to 1770, for block 745 */
    else if (!strcmp(priv_proto, "DEFAULT"))
    {
        const oid *p = get_default_privtype(&session.securityPrivProtoLen);
        session.securityPrivProto =
            snmp_duplicate_objid(p, session.securityPrivProtoLen);
    }
```

`session.securityEngineID` and `session.contextEngineID` are created in Net-SNMP tools.c:470 with `char           *s = calloc(1, olen ? olen : 1), *op = s;` and destroyed in Net-SNMP snmp_api.c:1928-1953:

```c
static void
snmp_free_session(netsnmp_session * s)
{
    if (s) {
        SNMP_FREE(s->localname);
        SNMP_FREE(s->peername);
        SNMP_FREE(s->community);
        SNMP_FREE(s->contextEngineID);
        SNMP_FREE(s->contextName);
        SNMP_FREE(s->securityEngineID);
        SNMP_FREE(s->securityName);
        SNMP_FREE(s->securityAuthProto);
        SNMP_FREE(s->securityPrivProto);
        SNMP_FREE(s->paramName);
#ifndef NETSNMP_NO_TRAP_STATS
        SNMP_FREE(s->trap_stats);
#endif /* NETSNMP_NO_TRAP_STATS */

        /*
         * clear session from any callbacks
         */
        netsnmp_callback_clear_client_arg(s, 0, 0);

        free((char *) s);
    }
}
```

I'll have to read up on `calloc/malloc` behaviors this week and see if we're passing something to the functions that violate their intended use cases.

**Reply by kamakazikamikaze** [reply]
Using documentation from the [Net-SNMP developer's guide](http://www.net-snmp.org/wiki/index.php/TUT:Simple_Application), it appears that some calls for dynamic value management are unnecessary. Replacing them by pointing to constants instead appears to plug most leaks. I'm just down to fixing the `hex_to_binary2` leaks.

**Comment by kamakazikamikaze** [comment]
I believe I've detected the source of the leaks for v3 sessions. It appears that calling `snmp_sess_open` eventually results in our session being cloned and the original left unfree'd, rather than having our session added to the list of internal sessions as it is. The calling order is as follows: `snmp_sess_open -> _sess_open -> snmp_sess_add -> snmp_sess_add_ex -> snmp_sess_copy -> _sess_copy`. The solution appears to be to free our `malloc` calls in our `create_session_capsule` function after we've opened the session.

**Comment by kamakazikamikaze** [comment]
Outstanding issues revolving around memory leaks have been addressed. This conversation will be locked for archival purposes and may be unlocked in the future if the issue(s) resurface.
